state/undertaker.go

```diff
@@ -25,14 +30,15 @@ func (st *State) ProcessDyingModel() (err error) {
 		}
 
 		if st.IsController() {
+			// We should not mark the controller model as Dead until
+			// all hosted models have been removed, otherwise the
+			// hosted model environs may not have beeen destroyed.
 			models, err := st.AllModels()
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
-			for _, model := range models {
-				if model.UUID() != st.ModelUUID() && model.Life() != Dead {
-					return nil, errors.Errorf("one or more hosted models are not yet dead")
-				}
+			if n := len(models) - 1; n > 0 {
```

Comment:
AllModels() call above will return even those models that are already Dead. Do we want them to be part of this count?